### ISSUE TITLE
`time spack spec visit` takes a really long time (over 6 hours)

### ISSUE THREAD

**Issue by goxberry** [spack/spack#7057]
Summary

Running `time spack spec visit` takes over 6 hours and counting:

```console
Uvalde:~ oxberry1$ time spack spec visit
Input spec
--------------------------------
visit

Concretized
--------------------------------
^C
==> Error: Keyboard interrupt.

real    416m9.002s
user    314m12.662s
sys     15m5.101s
```

### Expected Result

`time spack spec visit` returns in, say, 5 minutes or less.

### Actual Result

What happened instead

### Steps to reproduce the issue

```console
$ time spack spec visit
$ #wait 6 hours
```

### Information on your system

Platform: macOS/OS X Sierra, version 10.12.6, spack commit `b5daa51c06095fde57ffe05622bb1079355d3b73`

lmod modules loaded:
```console
Currently Loaded Modules:
  1) ncurses/6.0   2) zlib/1.2.11   3) emacs/25.3   4) openssl/1.0.2n   5) libevent/2.0.21   6) tmux/2.6   7) exuberant-ctags/5.8   8) global/6.5
```

output of `ml --version`:
```console
Modules based on Lua: Version 7.7.13  2017-10-31 10:46 -05:00
    by Robert McLay mclay@tacc.utexas.edu
```

packages.yaml:
```console
packages:
  perl:
    variants: +cpanm+shared
  ruby:
    variants: +openssl+readline
  pcre:
    variants: +jit+utf
  netcdf:
    variants: +mpi
  hdf5:
    variants: +mpi
  qt:
    variants: +opengl
  all:
    compiler: [clang@8.1.0-apple, clang, gcc]
    providers:
      awk: [gawk]
      blas: [openblas, atlas, intel-mkl, netlib-lapack]
      daal: [intel-daal]
      elf: [elfutils]
      golang: [gcc]
      ipp: [intel-ipp]
      java: [jdk]
      lapack: [openblas, atlas, intel-mkl, netlib-lapack]
      mkl: [intel-mkl]
      mpe: [mpe2]
      mpi: [openmpi, mpich, intel-mpi]
      opencl: [pocl]
      openfoam: [openfoam-com, openfoam-org, foam-extend]
      pil: [py-pillow]
      pkgconfig: [pkgconf, pkg-config]
      scalapack: [netlib-scalapack, intel-mkl]
      szip: [libszip, libaec]
      tbb: [intel-tbb]
      jpeg: [libjpeg-turbo, libjpeg]
      cmake: [cmake%clang]
    paths: {}
    modules: {}
    buildable: true
    version: []
```

compilers.yaml
```console
compilers:
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: sierra
    paths:
      cc: /usr/bin/clang
      cxx: /usr/bin/clang++
      f77: /Users/oxberry1/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/gcc-7.2.0-swfonpqzedbxtb6gfy7t4yphnw4vrkr2/bin/gfortran
      fc: /Users/oxberry1/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/gcc-7.2.0-swfonpqzedbxtb6gfy7t4yphnw4vrkr2/bin/gfortran
    spec: clang@8.1.0-apple
    target: x86_64
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: sierra
    paths:
      cc: /Users/oxberry1/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/gcc-7.2.0-swfonpqzedbxtb6gfy7t4yphnw4vrkr2/bin/gcc
      cxx: /Users/oxberry1/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/gcc-7.2.0-swfonpqzedbxtb6gfy7t4yphnw4vrkr2/bin/g++
      f77: /Users/oxberry1/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/gcc-7.2.0-swfonpqzedbxtb6gfy7t4yphnw4vrkr2/bin/gfortran
      fc: /Users/oxberry1/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/gcc-7.2.0-swfonpqzedbxtb6gfy7t4yphnw4vrkr2/bin/gfortran
    spec: gcc@7.2.0
    target: x86_64
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: sierra
    paths:
      cc: /Users/oxberry1/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/llvm-5.0.1-nadhxw7izy3thxuzopkinnqrecg76h2n/bin/clang
      cxx: /Users/oxberry1/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/llvm-5.0.1-nadhxw7izy3thxuzopkinnqrecg76h2n/bin/clang++
      f77: /Users/oxberry1/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/gcc-7.2.0-swfonpqzedbxtb6gfy7t4yphnw4vrkr2/bin/gfortran
      fc: /Users/oxberry1/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/gcc-7.2.0-swfonpqzedbxtb6gfy7t4yphnw4vrkr2/bin/gfortran
    spec: clang@5.0.1
    target: x86_6
```

modules.yaml
```console
modules:
  enable::
    - lmod
  lmod:
    core_compilers:
      - 'clang@8.1.0-apple'
    hierarchy:
      - mpi
      - lapack
    hash_length: 0
    ^python:
      autoload: 'direct'
    mumps:
      suffixes:
        '~mpi': no-mpi
    all:
      suffixes:
        '+debug': debug
```

**Comment by alalazo** [comment]
Semi-related thought: when looking at #6884 I was thinking that we could benefit from  more logging in the concretizer. Probably an option to enumerate and dump the graph being constructed at given points in the concretization algorithm could help debugging issues of this kind. Bonus if we could explore the dumps graphically and see visually what are the decisions of the concretizer.

**Comment by alalazo** [comment]
I can reproduce this (well, waited a few minutes, not 6 hours). If it may be of help, you can comment the `cmake` and `python` dependencies and still have long concretization times:
```
$ time spack spec visit
Input spec
--------------------------------
visit

Concretized
--------------------------------
^C
==> Error: Keyboard interrupt.

real	3m9.058s
user	3m7.668s
sys	0m1.324s
```
 Any further dependency eliminated reduces the concretization time to a few seconds. Using a default checkout (no custom configuration) on:
```console
$ uname -a
Linux nuvolari 4.4.0-109-generic #132~14.04.1-Ubuntu SMP Tue Jan 9 21:46:42 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by alalazo** [comment]
Further reduced to:
```console
$ spack spec vtk@6.1.0~opengl2+qt
```
Just a guess, but I think the `gl`  requirements in `vtk` are the root of this issue.

**Comment by scheibelp** [comment]
> Just a guess, but I think the gl requirements in vtk are the root of this issue.

That would make sense since https://github.com/spack/spack/pull/6551 was merged a couple days ago

**Comment by alalazo** [comment]
So, as far as I can see a package with just:
```python
depends_on('opengl', when='~osmesa')
depends_on('qt+opengl', when='+qt')
```
reproduces the issue.

**Comment by goxberry** [comment]
> Just a guess, but I think the `gl` requirements in `vtk` are the root of this issue.

That was my impression as well, because I've run into other issues with #6551 also, once I was able to add enough constraints to force concretization. In particular:

```console
Uvalde:build oxberry1$ spack spec visit^qt@4.8.6+opengl
Input spec
--------------------------------
visit
    ^qt@4.8.6+opengl

Concretized
--------------------------------
==> Error: There are no valid versions for spectrum-mpi that match ':'
```

is just bizarre, but `spack spec visit^qt@4.8.6+opengl^hdf5+mpi^openmpi` will concretize. The output of that last command is long enough that I'm not sure it's worth pasting.

Once I got the spec to concretize, I ran into install bugs with `vtk` dependencies added by #6551 (`libharu` doesn't build with my config). I don't have the bandwidth now to fix them, but if I get around to fixing it, I'll submit a PR.
